sys/uio.c

```diff
@@ -0,0 +1,88 @@
+/*
+ * Phoenix-RTOS
+ *
+ * libphoenix
+ *
+ * sys/uio.c
+ *
+ * Copyright 2021 Phoenix Systems
+ * Author: Ziemowit Leszczynski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <errno.h>
+#include <unistd.h>
+#include <limits.h>
+#include <sys/uio.h>
+
+
+static int checkv(const struct iovec *iov, int iovcnt)
+{
+	ssize_t max_len = SSIZE_MAX;
+
+	if (iovcnt <= 0 || iovcnt > IOV_MAX)
+		return -1;
+
+	for (unsigned int i = 0; i < iovcnt; ++i) {
+		if (iov[i].iov_len > max_len)
+			return -1;
+		max_len -= iov[i].iov_len;
+	}
+
+	return 0;
+}
+
+
+ssize_t readv(int fd, const struct iovec *iov, int iovcnt)
+{
+	ssize_t tot_len = 0;
+
+	if (-1 == checkv(iov, iovcnt))
+		return -EINVAL;
+
+	/* TODO: file locking */
+
+	for (unsigned int i = 0; i < iovcnt; ++i) {
+		if (iov[i].iov_len > 0) {
+			ssize_t n = read(fd, iov[i].iov_base, iov[i].iov_len);
```

Comment:
As above, variables should be defined before the function code.

Reply:
please note that it's false that C89 allows variable declaration at the beginning of the function - it actually allows declarations at the beginnig of each compound statement (commonly called a "block" - `{ .. }`) - exact syntax from C89 specification: http://port70.net/~nsz/c/c89/c89-draft.html#3.6.2.

If we're going for C89 compatibility, we should reflect that in our code guidelines and in CFLAGS (`-std=c89 -pedantic`) which will make non-compliant code throw compilation errors :).

IMHO, this code is a valid C89 and also limits the scope of the variable which improves the readability and reduces change for additional errors (by limiting the scope of variable).

Of course, declaring variables inside `for`/`while` (the above comment) is not C89-compliant.